add extra_flags attribute to pkl_project

asalah33 opened the pull request:
Should resolve #23

KushalP commented:
Can you add a test to flex this behaviour please.

sitaktif commented:
If we want to avoid spaces, we can `.strip()` the values, but then it makes it impossible for users to have values with spaces.

`repository_ctx.execute()` ([ref](https://bazel.build/rules/lib/builtins/repository_ctx#execute)) doesn't wrap the command in a shell afaik, so no need to escapes spaces.

asalah33 commented on pkl/extensions.bzl:
```diff
@@ -34,6 +34,21 @@ pkl_project = tag_class(
+                See the Pkl docs: https://pkl-lang.org/main/current/pkl-cli/index.html#command-eval""",
+            default = [],
+        ),
+        "environment": attr.string_dict(
```
Used `environment` (instead of `env`) to better align with the terminology in [Bazel's documentation on `repository_ctx.execute`](https://bazel.build/rules/lib/builtins/repository_ctx#execute).

asalah33 commented:
I’ve added tests and introduced `environment` and `properties` alongside `extra_flags`.
Let me know if you have any additional feedback!

cc @KushalP @sitaktif @holzensp

holzensp commented on pkl/private/pkl_project.bzl:
```diff
@@ -45,6 +45,24 @@ def parse_pkl_project_deps_json(pkl_deps_json_string_path):
+            - **Keys** are the names of the item (e.g., "key1", "key2").
+            - **Values** are the corresponding values for each item (e.g., "value1", "value2").
+    Returns:
+        A list of strings containing PKL CLI options.
```
```suggestion
        A list of strings containing Pkl CLI options.
```

holzensp commented on pkl/private/pkl_project.bzl:
```diff
@@ -45,6 +45,24 @@ def parse_pkl_project_deps_json(pkl_deps_json_string_path):
+    list = []
+    for key, value in items_dict.items():
+        list.append(option_name)
+        list.append("{key}={value}".format(key = key, value = value))
```
```suggestion
        list.append("'{key}'='{value}'".format(key = key, value = value))
```
See remark about test coverage

asalah33 replied:
I think adding quotation marks around key or value could be decided by the user based on thier needs
For example, a value like `I'm good` will break the evaluation command, as we surround it with single quotation marks

but if we left the decision to the user, it could be passed as `"I'm good"` and evaluation command would be happy

What you think?

holzensp commented on tests/integration_tests/example_workspaces/pkl_project/MODULE.bazel:
```diff
@@ -0,0 +1,59 @@
+
+    # Simpler way to pass environment variables during the evaluation of the PklProject file.
+    environment = { 
+        "PACKAGE_VERSION": "1.0.0",
```
Coverage could be slightly more extensive
```suggestion
        "PACKAGE_VERSION": "1.0.0",
        "RANDOM_VARIABLE": "Alternative '\"😀 content!!&;"
```

holzensp commented on tests/integration_tests/example_workspaces/pkl_project/PklProject:
```diff
@@ -0,0 +1,15 @@
+amends "pkl:Project"
+
+package {
+  name = read("env:PACKAGE_NAME")
```
Pkl's project evaluator doesn't load the `prop:` resource reader, even though `--property` is listed in e.g. `pkl project resolve --help`. Was there a specific reason to leave out the property reader, @bioball?

Either way; `prop:` is not supported in `PklProject` up to and including the current version (0.27), so it's probably best to leave out the `properties` attribute for now.

asalah33 replied:
Removed `properties` from this PR

holzensp reviewed (commented):
Almost there. I should have caught the `properties` before; currently, Pkl doesn't read `prop:` resources in `PklProject`.

For now, to land this, probably best to drop `properties`, but certainly extend the `environment` test cases a bit (I put in one suggestion that covers a few things).
